File: sections/32_combinatorics_under_modulo/problems/c_distribute_candies/solution.py
import sys


MOD = 1_000_000_007


def prepare(limit: int) -> tuple[list[int], list[int]]:
    fact = [1] * (limit + 1)
    for i in range(1, limit + 1):
        fact[i] = fact[i - 1] * i % MOD
    invfact = [1] * (limit + 1)
    invfact[limit] = pow(fact[limit], MOD - 2, MOD)
    for i in range(limit, 0, -1):
        invfact[i - 1] = invfact[i] * i % MOD
    return fact, invfact


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    queries = [(data[i], data[i + 1]) for i in range(1, 2 * q + 1, 2)]
    limit = max(n + s - 1 for n, s in queries)
    fact, invfact = prepare(limit)
    out = []
    for n, s in queries:
        total = n + s - 1
        choose = n - 1
        out.append(str(fact[total] * invfact[choose] % MOD * invfact[total - choose] % MOD))
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/32_combinatorics_under_modulo/problems/c_distribute_candies/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1000000007LL;

long long mod_pow(long long a, long long b) {
    long long result = 1;
    while (b > 0) {
        if (b & 1LL) result = result * a % MOD;
        a = a * a % MOD;
        b >>= 1LL;
    }
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<pair<int, int>> queries(q);
    int limit = 0;
    for (auto &[n, s] : queries) {
        cin >> n >> s;
        limit = max(limit, n + s - 1);
    }
    vector<long long> fact(limit + 1, 1), invfact(limit + 1, 1);
    for (int i = 1; i <= limit; i++) fact[i] = fact[i - 1] * i % MOD;
    invfact[limit] = mod_pow(fact[limit], MOD - 2);
    for (int i = limit; i >= 1; i--) invfact[i - 1] = invfact[i] * i % MOD;
    for (auto [n, s] : queries) {
        int total = n + s - 1;
        int choose = n - 1;
        cout << fact[total] * invfact[choose] % MOD * invfact[total - choose] % MOD << '\n';
    }
    return 0;
}
